### PR TITLE
[Relay][VM]VM Profiler

### PR DESCRIPTION
**Pull request by wweic**
Basic profiler functionality to get operator execution time.

Example output:
````bash
#OpName                       	#InvokeCount	#Duration(us): Sum/Mean/Min/Max
fused_nn_softmax              	1         	12.035/12.035/12.035/12.035
fused_nn_bias_add             	1         	1.149/1.149/1.149/1.149
fused_nn_dense                	1         	119.539/119.539/119.539/119.539
fused_nn_batch_flatten        	1         	5.081/5.081/5.081/5.081
fused_nn_global_avg_pool2d    	1         	11.423/11.423/11.423/11.423
fused_add_19                  	2         	73.225/36.6125/36.438/36.787
fused_nn_conv2d_11            	1         	391.84/391.84/391.84/391.84
fused_nn_conv2d_10            	3         	19419.7/6473.22/5625.45/8143.65
fused_nn_relu_4               	4         	217.312/54.328/30.015/102.374
fused_add_18                  	4         	164.761/41.1903/32.124/66.364
Total Duration 107960 us
````
cc @zhiics @icemelon9 @jroesch

### CODE REVIEW

File: src/runtime/vm/debug/debug_vm.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  void InvokePacked(Index packed_index, const PackedFunc& func, Index arg_count,
+                    Index output_size, const std::vector<Object>& args) final;
+
+  ~VirtualMachineDebug() {}
+
+ private:
+  void Init(const std::vector<TVMContext>& ctxs);
+
+  std::unordered_map<Index, std::string> packed_index_map;
+  std::unordered_map<Index, double> op_durations;
```
Comment (icemelon9) [inline src/runtime/vm/debug/debug_vm.h]: Use a vector to store op_durations, and report mean (also max/min like nvprof?) in the stats.

File: src/runtime/vm/debug/debug_vm.cc [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+void VirtualMachineDebug::InvokePacked(Index packed_index,
+                                       const PackedFunc& func, Index arg_count,
+                                       Index output_size,
+                                       const std::vector<Object>& args) {
+  auto op_begin = std::chrono::high_resolution_clock::now();
+  VirtualMachine::InvokePacked(packed_index, func, arg_count, output_size,
+                               args);
+  auto op_end = std::chrono::high_resolution_clock::now();
+  double op_duration =
+      std::chrono::duration_cast<std::chrono::duration<double> >(op_end -
```
Comment (icemelon9) [inline src/runtime/vm/debug/debug_vm.cc]: Use `std::chrono::microseconds` instead?

File: include/tvm/runtime/vm.h [hunk 3]
```diff
@@ -375,8 +375,41 @@ struct VMFrame {
...
+   *  executing the function on certain devices.
+   *  For benchmarking, use prepare to eliminate
+   *
+   * \param name the name of the function.
+   * \param sptr_to_self The shared_ptr that points to this module node.
+   *
+   * \return PackedFunc(nullptr) when it is not available.
+   *
+   * \note The function will always remain valid.
+   *   If the function need resource from the module(e.g. late linking),
```
Comment (zhiics) [inline include/tvm/runtime/vm.h]:
```suggestion
   *   If the function needs resource from the module(e.g. late linking),
```

File: python/tvm/relay/backend/debug_vm.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        for dev, tgt in target.items():
+            dev_type = tvm.expr.IntImm("int32", tvm.nd.context(dev).device_type)
+            tgts[dev_type] = tvm.target.create(tgt)
+    else:
+        raise TypeError("target is expected to be str, tvm.target.Target, " +
+                        "or dict of str to str/tvm.target.Target, but received " +
+                        "{}".format(type(target)))
+    return tgts
+
+class VMCompilerDebug(object):
```
Comment (zhiics) [inline python/tvm/relay/backend/debug_vm.py]: Can we inherit this class from `VMCompiler`?

File: src/runtime/vm/debug/debug_vm.cc (hunk 2 above)
Comment (zhiics) [reply src/runtime/vm/debug/debug_vm.cc]: I think high_resolution_clock is fine. But do we need to sync before measurement to profile devices other than CPU?

File: src/runtime/vm/debug/debug_vm.h (hunk 1 above)
Comment (wweic) [reply src/runtime/vm/debug/debug_vm.h]: done

File: src/runtime/vm/debug/debug_vm.cc (hunk 2 above)
Comment (wweic) [reply src/runtime/vm/debug/debug_vm.cc]: maybe I add a todo here once we have heterogeneous execution? https://github.com/dmlc/tvm/blob/master/src/runtime/graph/debug/graph_runtime_debug.cc#L80

File: CMakeLists.txt [hunk 5]
```diff
@@ -198,6 +198,12 @@ if(USE_GRAPH_RUNTIME)
   endif(USE_GRAPH_RUNTIME_DEBUG)
 endif(USE_GRAPH_RUNTIME)
 
+message(STATUS "Build with Relay VM debug support...")
```
[the remark this answers is not in the document]
Comment (wweic) [reply CMakeLists.txt]: done

**Comment by u99127** [comment]
A higher level comment is that such functionality is usually referred to as profiling rather than debugging. What do others think ?

**Comment by wweic** [comment]
> A higher level comment is that such functionality is usually referred to as profiling rather than debugging. What do others think ?

Agree what I added lies more in profiling. I'm following the name for current graph runtime debugger(https://docs.tvm.ai/dev/debugger.html), which supports inspecting intermediate tensors and operator execution metrics. I'm basically replicating the behavior for Relay VM hence the name. I expect eventually it will be more like a debugger than a profiler.

**Comment by u99127** [comment]
> > A higher level comment is that such functionality is usually referred to as profiling rather than debugging. What do others think ?
> 
> Agree what I added lies more in profiling. I'm following the name for current graph runtime debugger(https://docs.tvm.ai/dev/debugger.html), which supports inspecting intermediate tensors and operator execution metrics. I'm basically replicating the behavior for Relay VM hence the name. I expect eventually it will be more like a debugger than a profiler.

Ok, makes sense now. Thanks.

File: src/runtime/vm/debug/debug_vm.cc (hunk 2 above)
Comment (icemelon9) [reply src/runtime/vm/debug/debug_vm.cc]: Since VM now supports running on CUDA, we should probably handle this here.

Comment (wweic) [reply src/runtime/vm/debug/debug_vm.cc]: thanks. done.

**Comment by icemelon9** [comment]
You should also set USE_VM_DEBUG to OFF by default in the config.cmake.

**Comment by jroesch** [comment]
I also agree with @u99127 I don't know why the TVM runtime calls this a debugger, this is way more in profiling land then debugger land. 

I would imagine a VM debugger to be a typical step debugger instead of just profiling info.

File: python/tvm/relay/backend/debug_vm.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        for dev, tgt in target.items():
+            dev_type = tvm.expr.IntImm("int32", tvm.nd.context(dev).device_type)
+            tgts[dev_type] = tvm.target.create(tgt)
+    else:
+        raise TypeError("target is expected to be str, tvm.target.Target, " +
+                        "or dict of str to str/tvm.target.Target, but received " +
+                        "{}".format(type(target)))
+    return tgts
+
+class VMCompilerDebug(vm.VMCompiler):
```
Comment (jroesch) [inline python/tvm/relay/backend/debug_vm.py]: Do we actually need a different entry point for the debugging the VM? can't we just pass an option into the normal build target?

**Comment by zhiics** [comment]
Yeah, we probably can rename it to profiler since debugger is different than what we are doing.

A side node that can be probably a follow-up pr is that maybe we can add memory stats for each op as well through memory_manager or RSS.

**Comment by yongwww** [comment]
+ to call this a profiler.

File: python/tvm/relay/backend/debug_vm.py (hunk 6 above)
Comment (wweic) [reply python/tvm/relay/backend/debug_vm.py]: If we want to pass in an option, then `VMCompiler` will need to switch between `VirtualMachine` and `VirtualMachineDebug` based on the config, which means we have to build `VirtualMachineDebug` all the time. I think we need the option to not build the `VirtualMachineDebug` source, does it make sense?

**Comment by wweic** [comment]
Thanks. I renamed the PR to profiler. Luckily I don't think we need to change sources since I haven't used the term `debugger` in sources. But let me know if you have opinions about file naming.

**Comment by zhiics** [comment]
Do we need to change the file name and class name as well?

File: src/runtime/vm/profiler/vm.cc [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file src/runtime/vm/debug/vm_debug.cc
```
Comment (zhiics) [inline src/runtime/vm/profiler/vm.cc]: we probably should change the class name and the \file directory and name as well to keep them consistent to the Python side. Please also check profiler/vm.h

**Pull request merged**